### DISCUSSION TITLE
How to complete the OAuth flow if user with that email doesn't exist

### DISCUSSION THREAD

**Discussion by nikolaydyankov** [pocketbase/pocketbase#2471] [Q&A] [answered]
First of all - thank you so much for developing this project, I'm a big fanboy of PocketBase.

I'm authenticating client-side like this:

```javascript
const pb = new PocketBase('http://127.0.0.1:8090')
const authMethods = await pb.collection('users').listAuthMethods()
const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' })
```

It works fine when there is already a user with the provided email in the database, but if there isn't, this is what I see in the console:

```
POST http://127.0.0.1:8090/api/collections/users/auth-with-oauth2 400 (Bad Request)

ClientResponseError 400: Failed to authenticate.
    at new ClientResponseError (ClientResponseError.ts:13:9)
    at RecordService.<anonymous> (RecordService.ts:472:32)
    at step (tslib.es6.js:147:23)
    at Object.throw (tslib.es6.js:128:53)
    at rejected (tslib.es6.js:119:64)
```

What should I do in this situation? If the user doesn't exist in the database I would still like to get the email from the OAuth flow and create a new user (with auto generated password? this part is not very clear).

**Comment by ganigeorgiev** [comment]
If user with the OAuth2 account email doesn't exist, then PocketBase will make an attempt to create it.

I guess it can fail if your `users` collection have custom required fields. You can pass additional user create data with the OAuth2 call using the `createData` prop:

```js
await pb.collection('users').authWithOAuth2({
    provider: 'google',
    createData: {
        yourField: "example"
    }
})
```

(_file uploads or referencing OAuth2 raw user data fields are currently not supported via the `createData` prop_)

You can `try/catch` and inspect the returned `ClientResponseError` to check for the error fields (eg. `err.response.data`) and/or check the failed request log in the Admin UI > Logs.

**Comment by nikolaydyankov** [comment]
This is the error from the admin logs:

```
{
  "errorDetails": "constraint failed: UNIQUE constraint failed: _externalAuths.collectionId, _externalAuths.recordId, _externalAuths.provider (2067)",
  "errorMessage": "Failed to authenticate."
}
```

And the `ClientResponseError`:

```
{
  "url": "http://127.0.0.1:8090/api/collections/users/auth-with-oauth2",
  "status": 400,
  "response": {
    "code": 400,
    "message": "Failed to authenticate.",
    "data": {}
  },
  "isAbort": false,
  "originalError": null,
  "name": "ClientResponseError 400"
}
```

I tried setting the custom fields with `createData`, but no luck. I also don't have any unique constraints.

**Reply by ganigeorgiev** [reply]
Hm, that's strange. Could you provide a more complete code sample how do you authenticate?

Before calling `authWithOAuth2` do you have an existing auth model in `pb.authStore` (aka. check if `pb.authStore.token` is empty)?

If you have the `sqlite3` cli tool installed, could you inspect the content of the `_externalAuths` table (eg. `sqlite3 ./pb_data/data.db 'select * from _externalAuths'`)?
Note that currently a single user can be linked with only one OAuth2 account from a single OAuth2 provider.

**Comment by nikolaydyankov** [comment, marked as answer]
I added an `authStore.clear()` before the auth call and it worked. I commented it out and it's still working. I guess there was some state stuck in the auth table?

```javascript
async function signInWithGoogle() {
    try {
      const pb = new PocketBase('http://127.0.0.1:8090')
      pb.authStore.clear()
      const authMethods = await pb.collection('users').listAuthMethods()
      const authData = await pb.collection('users').authWithOAuth2({
        provider: 'google',
        // todo
        createData: {
          subscribed: false,
          billing_date: null,
          stripe_customer_id: null,
        },
      })
      console.log(authData)
      console.log(pb.authStore.exportToCookie({ httpOnly: false }))
    } catch (e) {
      console.log(e.response)
    }
  }
```

```html
<Button text="Continue with Google" onClick={signInWithGoogle} {loading} />
```

If you don't think this is my error I will install the sqlite3 tool and try to give you the info.

**Reply by ganigeorgiev** [reply]
> I added an authStore.clear() before the auth call and it worked. I commented it out and it's still working. I guess there was some state stuck in the auth table?

If clearing the authStore has fixed the issue, then I suspect that you were previously authenticated with a user that was already linked to a another Google account. This is expected because with the OAuth2 auth call we also send the `Authorization` header to allow manually linking (eg. in case the account was previously created with email/pass or the OAuth2 provider didn't return an email, or it is different from the user one). For example:

```js
// authenticate with password (or as in your case probably loaded from a previous localStorage state)
await pb.collection("user").authWithPassword("test@example.com", "123456");

// will send the previously authenticated user token with the request and will link the OAuth2 user (if not already associated to another PocketBase user)
await pb.collection("user").authWithOAuth2({ provider: "google" });
```

---

> Quick side question - I can't seem to get import PocketBase from 'pocketbase' working in a SvelteKit component.

What is the error that is returned? Note that if you have the PocketBase executable alongside your SvelteKit files and depending how your bundler is configured, it is possible that it may make an attempt to load the local "pocketbase" file instead looking in the `node_modules` directory for the js package. If that's the case, try to move the binary (eg. in a sub directory) to avoid the import conflict.

**Reply by nikolaydyankov** [reply]
I'm guessing something got stuck in the localStorage, because I normally do server-side auth with `restoreAuth(cookie)` and don't touch local storage at all.

> What is the error that is returned? Note that if you have the PocketBase executable alongside your SvelteKit files and depending how your bundler is configured, it is possible that it may make an attempt to load the local "pocketbase" file instead looking in the node_modules directory for the js package. If that's the case, try to move the binary (eg. in a sub directory) to avoid the import conflict.

That was the issue, thanks.

**Reply by ganigeorgiev** [reply]
> I normally do server-side auth with restoreAuth(cookie) and don't touch local storage at all.

The default PocketBase auth store is using localStorage when initialized in the browser so probably something like this has happened:

1. Called `signInWithGoogle()` with Google account A.
   _This will automatically persist the auth model and token in the localStorage._

3. Refreshed the page and decided to call again `signInWithGoogle()` with Google account B.
   _On `new PocketBase(...)` inside the browser this will automatically load the existing localStorage state and effectively authenticating you with the previously created PocketBase user._

**Comment by hungcrush** [comment]
@ganigeorgiev 

> If user with the OAuth2 account email doesn't exist, then PocketBase will make an attempt to create it.

How can I know the user is newest? I need to show the dialog and ask user update more information, if that's first time..

**Reply by ganigeorgiev** [reply]
The JSON auth response currently doesn't contain information whether the user was just created or it is existing (this will be added in the [next v0.16/v0.17 release](https://github.com/orgs/pocketbase/projects/2?pane=issue&itemId=28175380)).

If you use PocketBase as framework, you can attach to the [`OnRecordAfterAuthWithOAuth2Request` hook](https://pocketbase.io/docs/event-hooks/#record-auth-api-hooks) and check its `e.IsCreate` field.

Alternatively, you can also determine whether the user is "new" by manually comparing its created/updated fields (_if both date fields are equal then the user didn't have a previous update operation_)

**Reply by hungcrush** [reply]
Thank you. I got it. Now i'm checking if record.name === "" ==> it's newest 😬  Why don't you use the `name` field in the OAuth data when creating account?

**Reply by ganigeorgiev** [reply]
> Why don't you use the name field in the OAuth data when creating account?

Because the auth collection may not have a `name` field or it could be named something differently (eg. `fullName`, "firstName" + "lastName", etc.).

**Comment by RajaRakoto** [comment]
I recently encountered a similar issue to the one described in the provided link, and I am seeking assistance in resolving it. If anyone can help me, it would be greatly appreciated.

[OAuth authentication issue with authWithOAuth2(): Bad Request error](https://github.com/pocketbase/pocketbase/issues/2905)
